actions/show-heartbeat/index.js

```diff
@@ -1,20 +1,90 @@
-/* global Normandy */
-import uuid from 'uuid';
-
-
-Normandy.registerAction('show-heartbeat', function(_, args) {
-    let flowId = uuid.v4();
-
-    // A bit redundant but the action argument names shouldn't necessarily rely
-    // on the argument names showHeartbeat takes.
-    Normandy.showHeartbeat({
-        message: args.message,
-        thanksMessage: args.thanksMessage,
-        flowId: flowId,
-        postAnswerUrl: args.postAnswerUrl,
-        learnMoreMessage: args.learnMoreMessage,
-        learnMoreUrl: args.learnMoreUrl,
-    }).then(() => {
-        console.log('Heartbeat happened!');
-    });
-});
+import 'babel-polyfill';
+
+import {Action, registerAction, weightedChoose} from '../utils';
+
+const VERSION = 52; // Increase when changed.
+const LAST_SHOWN_DELAY = 100 * 60 * 60 * 24 * 7; // 7 days
+
+export default class ShowHeartbeatAction extends Action {
+    constructor(normandy, recipe) {
+        super(normandy, recipe);
+        this.storage = normandy.createStorage(recipe.id);
+    }
+
+    async execute() {
+        let {surveys, defaults} = this.recipe.args;
+
+        let lastShown = await this.getLastShownDate();
+        if (lastShown !== null && Date.now() - lastShown < LAST_SHOWN_DELAY) {
+            return;
+        }
+
+        let survey = this.chooseSurvey(surveys, defaults);
+        let flowId = this.normandy.uuid();
+
+        // A bit redundant but the action argument names shouldn't necessarily rely
+        // on the argument names showHeartbeat takes.
+        await this.normandy.showHeartbeat({
+            message: survey.message,
+            thanksMessage: survey.thanksMessage,
+            flowId: flowId,
+            postAnswerUrl: await this.annotatePostAnswerUrl(survey.postAnswerUrl),
+            learnMoreMessage: survey.learnMoreMessage,
+            learnMoreUrl: survey.learnMoreUrl,
+        });
+
+        this.setLastShownDate();
+        this.normandy.log('Heartbeat happened!');
+    }
+
+    setLastShownDate() {
+        // Returns a promise, but there's nothing to do if it fails.
+        this.storage.setItem('lastShown', Date.now());
+    }
+
+    async getLastShownDate() {
+        let lastShown = Number.parseInt(await this.storage.getItem('lastShown'), 10);
+        return Number.isNaN(lastShown) ? null : lastShown;
+    }
+
+    async annotatePostAnswerUrl(url) {
+        let appInfo = await this.normandy.getAppInfo();
+        let params = (
+            'source=heartbeat' +
+            '&surveyversion=' + VERSION +
+            '&updateChannel=' + appInfo.defaultUpdateChannel +
+            '&fxVersion=' + appInfo.version
+        );
```

Comment:
I _really_ wish jsdom supported `URLUtils.searchParams`. I hate that we can't use a feature available in Firefox due to it not being mocked in our tests, but I don't think we have time to solve that right now.